File: apps/hrms-auth/src/main/java/com/cre/hrms/auth/controller/UserController.java
package com.cre.hrms.auth.controller;

import com.cre.hrms.security.authorization.SecurityUtils;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.HashMap;
import java.util.Map;

@RestController
@RequestMapping("/api/v1/users")
@RequiredArgsConstructor
public class UserController {

    @GetMapping("/me")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<Map<String, Object>> getCurrentUser() {
        Map<String, Object> response = new HashMap<>();
        SecurityUtils.getCurrentUsername().ifPresent(username -> {
            response.put("username", username);
            response.put("roles", SecurityUtils.getCurrentUserRoles());
        });
        return ResponseEntity.ok(response);
    }

    @GetMapping("/admin")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<String> adminEndpoint() {
        return ResponseEntity.ok("This is an admin-only endpoint");
    }

    @GetMapping("/hr")
    @PreAuthorize("hasAnyRole('HR', 'ADMIN')")
    public ResponseEntity<String> hrEndpoint() {
        return ResponseEntity.ok("This is an HR or Admin endpoint");
    }

    @GetMapping("/manager")
    @PreAuthorize("hasAnyRole('MANAGER', 'ADMIN')")
    public ResponseEntity<String> managerEndpoint() {
        return ResponseEntity.ok("This is a Manager or Admin endpoint");
    }
}

File: apps/hrms-auth/src/main/kotlin/com/cre/hrms/auth/controller/UserController.kt
package com.cre.hrms.auth.controller

import com.cre.hrms.security.authorization.SecurityUtils
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/v1/users")
class UserController {

    @GetMapping("/me")
    @PreAuthorize("isAuthenticated()")
    fun getCurrentUser(): ResponseEntity<Map<String, Any>> {
        val response = mutableMapOf<String, Any>()
        SecurityUtils.getCurrentUsername().ifPresent { username ->
            response["username"] = username
            response["roles"] = SecurityUtils.getCurrentUserRoles()
        }
        return ResponseEntity.ok(response)
    }

    @GetMapping("/admin")
    @PreAuthorize("hasRole('ADMIN')")
    fun adminEndpoint(): ResponseEntity<String> {
        return ResponseEntity.ok("This is an admin-only endpoint")
    }

    @GetMapping("/hr")
    @PreAuthorize("hasAnyRole('HR', 'ADMIN')")
    fun hrEndpoint(): ResponseEntity<String> {
        return ResponseEntity.ok("This is an HR or Admin endpoint")
    }

    @GetMapping("/manager")
    @PreAuthorize("hasAnyRole('MANAGER', 'ADMIN')")
    fun managerEndpoint(): ResponseEntity<String> {
        return ResponseEntity.ok("This is a Manager or Admin endpoint")
    }
}
